### PR TITLE
Change user flow for replying to a comment

### PR DESCRIPTION
**Pull request by AnkushMalik**
Now, for replying to a comment,
- Instead of directing the user to a different page(comment#new), a reply modal is opened for the user to write the reply on the same page(project#show).

Inspired from _Twitter_

peek 2018-06-02 23-27

### CODE REVIEW

File: app/views/comments/_comment.html.haml [hunk 1]
```diff
@@ -8,7 +8,21 @@
       = link_to user_path(comment.commenter) do
         = comment.commenter.name
       |
-      = link_to 'Reply', new_comment_comment_path( comment )
+      = link_to 'Reply', '#', "data-toggle" => "modal", 'data-target' => "#replyModal_#{comment.id}"
```
Comment (DavidKang) [inline app/views/comments/_comment.html.haml]: This could be written as: 
``` linkt_to('Reply', '#', data: { toggle: 'modal', target: "#replyModal_#{comment.id}") ```

File: app/views/comments/_comment.html.haml [hunk 2]
```diff
@@ -8,7 +8,21 @@
...
+      .modal.fade{id: "replyModal_#{comment.id}","aria-labelledby" => "replyModalLabel", role: "dialog", tabindex: "-1"}
```
Comment (DavidKang) [inline app/views/comments/_comment.html.haml]: Please use single-quotes instead of double-quotes for strings.

File: app/views/comments/_comment.html.haml [hunk 3]
```diff
@@ -8,7 +8,21 @@
...
+        .modal-dialog{role: "document"}
```
Comment (DavidKang) [inline app/views/comments/_comment.html.haml]: same here.

File: app/views/comments/_comment.html.haml [hunk 4]
```diff
@@ -8,7 +8,21 @@
...
+          .modal-content
+            .modal-header
+              %button.close{"aria-label" => "Close", "data-dismiss" => "modal", type: "button"}
```
Comment (DavidKang) [inline app/views/comments/_comment.html.haml]: same here.

File: app/views/comments/_form.html.haml [hunk 5]
```diff
@@ -1,23 +1,22 @@
...
+= render :partial => 'comments/help', locals: { unique_id: unique_id}
 = form_for [ parent, comment ], html: {role: "form"} do |f|
   %p
     .comment-form-heading
       %ul.nav.nav-tabs
-        %li.active#show-source{ role: :presentation }
-          %a{ href: '#markdown-source', role: :tab, data: { toggle: :tab } } Edit
-        %li#show-preview{ role: :presentation }
-          %a{ href: '#markdown-preview', role: :tab, data: { toggle: :tab } } Preview
+        %li.active{ id: "show-source#{unique_id}", role: :presentation }
```
Comment (DavidKang) [inline app/views/comments/_form.html.haml]: This could be `id` instead of `unique_id`, id should be unique. :smile_cat:

File: app/views/comments/_help.html.haml [hunk 6]
```diff
@@ -1,4 +1,5 @@
-#help{:class=>"collapse"}
+- unique_id = '' if unique_id == nil
```
Comment (DavidKang) [inline app/views/comments/_help.html.haml]: Why is this needed here?

File: app/views/comments/_help.html.haml [hunk 7]
```diff
@@ -22,6 +23,6 @@
   %p.text-right{:style=>"margin-bottom: 10px;"}
     = link_to "More Complex Markdown Help", "http://daringfireball.net/projects/markdown/syntax", target: "_blank"
 %p.text-right
-  %a{:type=>"button", "data-toggle"=>"collapse", "data-target"=>"#help", :class=>"btn btn-default btn-xs"}
+  %a{:type=>"button", "data-toggle"=>"collapse", "data-target"=>"#help#{unique_id}", :class=>"btn btn-default btn-xs"}
```
Comment (DavidKang) [inline app/views/comments/_help.html.haml]: Please use `:` instead of `=>`

File: app/views/comments/_help.html.haml (hunk 6 above)
Comment (AnkushMalik) [reply app/views/comments/_help.html.haml]: for project's form since theres no id in project form

File: app/views/comments/_form.html.haml (hunk 5 above)
Comment (AnkushMalik) [reply app/views/comments/_form.html.haml]: I tried to keep it different from the word 'id', as I thought it might be confusing for other devs

File: app/views/comments/_help.html.haml (hunk 6 above)
Comment (DavidKang) [reply app/views/comments/_help.html.haml]: Yes, but if is `nil` it will be transformed into an empty string.

Comment (AnkushMalik) [reply app/views/comments/_help.html.haml]: i put this condition because of "undefined local variable or method `unique_id' " since I didnt defined unique_id
i will define it as the local variable while rendering comments/help (like done in comment)

File: app/views/comments/_help.html.haml (hunk 7 above)
Comment (AnkushMalik) [reply app/views/comments/_help.html.haml]: done.
Can you please check why my Travis is showing failure in test cases? :/ there's not a single test failure on my local machine.

File: app/views/comments/_comment.html.haml [hunk 8]
```diff
@@ -8,10 +8,24 @@
       = link_to user_path(comment.commenter) do
         = comment.commenter.name
       |
-      = link_to 'Reply', new_comment_comment_path( comment )
+      = link_to('Reply', '#', data: { toggle: 'modal', target: "#replyModal_#{comment.id}" })
+      .modal.fade{id: "replyModal_#{comment.id}",'aria-labelledby' => 'replyModalLabel', role: 'dialog', tabindex: '-1'}
```
Comment (DavidKang) [inline app/views/comments/_comment.html.haml]: I don't see the point of rendering a `modal` per reply link. Maybe this should be moved to an action in the controller and call it via AJAX :thinking: . What do you think?

File: app/views/comments/_form.html.haml [hunk 9]
```diff
@@ -1,23 +1,22 @@
...
+= render :partial => 'comments/help', locals: { id: id}
 = form_for [ parent, comment ], html: {role: "form"} do |f|
   %p
     .comment-form-heading
       %ul.nav.nav-tabs
-        %li.active#show-source{ role: :presentation }
-          %a{ href: '#markdown-source', role: :tab, data: { toggle: :tab } } Edit
-        %li#show-preview{ role: :presentation }
-          %a{ href: '#markdown-preview', role: :tab, data: { toggle: :tab } } Preview
+        %li.active{ id: "show-source#{id}", role: :presentation }
```
Comment (DavidKang) [inline app/views/comments/_form.html.haml]: Why do you need all this `ids`, maybe because you are using `id` instead of `class`? if `show-source`, `show-preview` was a class you shouldn't need the random `id`. Even in the [js](https://github.com/SUSE/hackweek/pull/237/files#diff-a9c3bd311eab80c9ebe6a69830f9ad02R55) you are no using the `ids` you go to the form and after to the element you need to toggle.

File: app/views/comments/_comment.html.haml (hunk 8 above)
Comment (AnkushMalik) [reply app/views/comments/_comment.html.haml]: yeah, this approach will be better 😄 👍

Comment (AnkushMalik) [reply app/views/comments/_comment.html.haml]: actually theres no need of ajax as well i could just change the reply form's action everytime a reply btn is clicked with data attribute for which reply is clicked

File: app/controllers/comments_controller.rb [hunk 10]
```diff
@@ -21,6 +22,19 @@ def comment_params
...
+    @comment = Comment.find(params[:id])
+    @content = @comment.text.to_str.gsub(/:([\w+-]+):/) do |match|
+                 %(![add-emoji](https://assets-cdn.github.com/images/icons/emoji/#{match.to_str.tr(':','')}.png))
+               end if @comment.text
+    @rendered = MarkdownHelper.render @content
+    respond_to do |format|
+      format.html
+      format.js
+    end
+
```
Comment (DavidKang) [inline app/controllers/comments_controller.rb]: Remove this empty line

File: app/views/markdown/preview.js.erb [hunk 11]
```diff
@@ -1,4 +1,13 @@
-$('#preview-contents').html("<%=j raw @rendered %>");
-$('#loading-spinner').addClass('hidden');
-$('#preview-contents').removeClass('hidden');
+// ele = $('li[data-form-type="<%=params[:form_type]%>"]').closest('form');
```
Comment (DavidKang) [inline app/views/markdown/preview.js.erb]: Please, remove all the comments

Comment (AnkushMalik) [reply app/views/markdown/preview.js.erb]: done 😅
